### PR TITLE
Yakshaving the backend on my way to the science of data

### PR DESCRIPTION
**Pull request by 552020**
- [ ] Understand for what we need an internal session in our database.
- [ ] Understand if for the case we need an internal session, if it needs to be stored in the database or if it could be saved just in-memory.
- [ ] Consider the `with` keyword instead of closing manually the session in the `finally`. Check Note on Using SQLAlchemy objects with the `with` Statement
- [x] Do a bunch of other stuff ... (mostly just renaming, only 1 logical change)
- [x] Create a `full_text` and a `sanitized_text` columnd in the Chat model
- [x] Run the migration locally
- [ ] Remember to run the migration on the prod DB when this branch has been merged `alembic upgrade head
` ⚠️ 
- [x] Move full_text and stuff to their own table
- [x] Autogenerate alembic migration. Mind that the auto-generate migration doesn't work perfectly and there are well known cases where the chages are not reported, so it is always necessary to check it manualy https://alembic.sqlalchemy.org/en/latest/autogenerate.html#what-does-autogenerate-detect-and-what-does-it-not-detect
- [x] Fix automatically created migration, and apply them to the local database. Btw. it was a real hassle to sync it with the production one. 
- [ ] Save the chat to which user agreed to sell into the the full_text column in the ChatFullText table.
- [ ] To agree the last user needs to login, otherwise we will not have access to the Chat. The 'invitee' could not have an open session in Telegram, and if they agree to sell the chat, we need them to let us access Telegram to retrieve the chat. Probably is it more important to let them agree and then ask them as well any other user to login, so that we can retrieve the chat. This is UX problematic. It can be solved both ways. @dantol29 & @lmangall 
- [ ] Install cryptg (if not already installed), suggested in the docs: check paragragh below "About cryptg". I checked in the requirements.txt and I couldn't find it.
- [x] Tell @dantol29 about Telethon's chat: the official Telethon Telegram group: https://web.telegram.org/a/#-1001109500936. The developer of Telethon 'Ionami' is actively answering there. 
- [ ] Think if we need the user identifier in /add-user-to-agreed
- [x] Add logic to fetch the chat and save it as full_text in the database in the /add-user-to-agreed endpoint. A new 'status' has been added `fetched`. If a chat is `sold` but is not `fetched` this means that the last user who agreed didn't have an open session in Telethon/Telegram. It means the chat is sold but we still need to fetch it. The response is still 200.
- [x] Merge main into branch after rebase
- [x] Merge main into branch after adding black as autoformatter
- [x] Move `check_session_expiration` in its own file
- [x] Move commented out code for JWT and /access enpoint into routes/auth.py 
- [ ] Introduce the `logging` library/facility for Python: https://docs.python.org/3/library/logging.html. It will substitute our print calls. It has the same purpose of the Debug class in our WebServ. Quart has an app.logger built-in, but it's prefereable to work directly with the logging library. https://quart.palletsprojects.com/en/latest/how_to_guides/logging.html
- [x] switch from manual session handling to automatic session handling with the new function `get_sqlalchemy_session`. See below for an explanation. And update all the functions that were calling Session(). 
- [x] fix get-user endpoint: `"error": "Instance <User at 0x10d173410> is not bound to a Session; attribute refresh operation cannot proceed (Background on this error at: https://sqlalche.me/e/20/bhk3)"`
- [x] test /get-user
- [x] test /get-users
- [x] test /send-code
- [x] test /login
- [x] spend many hours to reinstall the whole postgres on my local machine on Sunday cause suddenly postgres is not working anymore. 
- [x] test /send-message
- [x] test /add-user-to-agreed

### **About cryptg**

> If [cryptg](https://github.com/cher-nov/cryptg) is installed, the library will work a lot faster, since encryption and decryption will be made in C instead of Python. If your code deals with a lot of updates or you are downloading/uploading a lot of files, you will notice a considerable speed-up (from a hundred kilobytes per second to several megabytes per second, if your connection allows it). If it’s not installed, [pyaes](https://github.com/ricmoo/pyaes) will be used (which is pure Python, so it’s much slower).

https://docs.telethon.dev/en/stable/basic/installation.html#optional-dependencies

---
### **Note on Using SQLAlchemy with the `with` Statement**

SQLAlchemy session objects are designed to work seamlessly with Python's `with` statement, which is used for context management. The `with` statement ensures that resources are properly managed by automatically invoking setup and cleanup operations.

When using a SQLAlchemy session within a `with` statement:
- **`__enter__` Method**: Sets up the session and returns the session object for use within the block.
- **`__exit__` Method**: Ensures the session is closed and any transactions are handled appropriately, even if an exception occurs.

This integration allows for cleaner and more maintainable code by automating resource management and reducing the risk of resource leaks. By leveraging context managers, SQLAlchemy simplifies the lifecycle management of database sessions, making it easier to ensure that sessions are properly closed after use.

---

### **Database Session Management Changes**

In our recent code refactor, we have updated how we handle database sessions to improve consistency, error handling, and resource management. Below is an explanation of what we were doing before and what we are doing now.

#### Previous Approach

**Imports and Session Creation:**
```python
from db import Session as DBSession

# Later in the code
db_session = DBSession()
```

**What We Were Doing:**
- We were importing `Session` from `db.py` and aliasing it as `DBSession`.
- `DBSession` was used to create a new session object each time we needed to interact with the database.
- The session was manually managed, including committing and closing the session.

**Example Function:**
```python
async def create_session(client, number, phone_hash, userId):
    session_id = client.session.save()
    db_session = DBSession()
    exit_code = 0
    try:
        if userId is None:
            userId = "None"
        new_session = Session(
            id=session_id,
            phone_number=number,
            phone_code_hash=phone_hash,
            user_id=userId,
        )
        db_session.add(new_session)
        db_session.commit()
        print(f"Creating new session for {number}({userId})")
    except Exception as e:
        print(f"Error creating new session: {str(e)}")
        exit_code = 1
    finally:
        db_session.close()
        return exit_code
```

**Issues with Previous Approach:**
- **Repetitive Code**: Manual handling of session creation, committing, and closing was repetitive.
- **Error Handling**: Manual error handling and rolling back transactions added complexity.
- **Resource Management**: Potential for forgetting to close the session, leading to resource leaks.

#### New Approach

**Imports and Session Creation:**
```python
from db import get_sqlalchemy_session

# Later in the code
async with get_sqlalchemy_session() as db_session:
```

**What We Are Doing Now:**
- We are using a context manager, `get_sqlalchemy_session`, from `db.py`.
- The context manager automatically handles session creation, committing, and closing.
- It simplifies error handling and ensures resources are properly managed.

**Example Function:**
```python
from db import get_sqlalchemy_session
import logging

logger = logging.getLogger(__name__)

async def create_session(client, number, phone_hash, userId):
    session_id = client.session.save()
    async with get_sqlalchemy_session() as db_session:
        exit_code = 0
        try:
            if userId is None:
                userId = "None"
            new_session = Session(
                id=session_id,
                phone_number=number,
                phone_code_hash=phone_hash,
                user_id=userId,
            )
            db_session.add(new_session)
            db_session.commit()
            logger.info(f"Creating new session for {number}({userId})")
        except Exception as e:
            logger.error(f"Error creating new session: {str(e)}")
            exit_code = 1
        return exit_code
```

**Benefits of New Approach:**
- **Consistency**: Ensures consistent session management across the application.
- **Error Handling**: Automatically handles commits, rollbacks, and logging errors.
- **Resource Management**: Guarantees that sessions are closed properly, preventing resource leaks.
- **Cleaner Code**: Reduces boilerplate code, making it easier to read and maintain.

### Summary

- **Old Method**: Manual session handling using `DBSession()`.
- **New Method**: Context manager `get_sqlalchemy_session` for automatic session handling.
- **Advantages**: Improved consistency, error handling, resource management, and code readability.

This change makes our codebase more robust and maintainable, ensuring better practices for database session management.

---

### CODE REVIEW

**Comment by dantol29** [comment]
How is this connected with data science. Looks like yak-shaving so far

**Comment by lmangall** [comment]
@dantol29 he's _on his way_ to data science, not there yet.
Maybe you should watch out he's going to catch up

**Comment by dantol29** [comment]
@552020 I would prefer not to safe text in the chat table. Chat tables are constantly being loaded, and there is no need to keep text in those. It may cause performance issues. Highly likely, we gonna use full text once to transform it to a sellable product. Therefore, it makes sense to make another table that not gonna be invoked every time we call /get-user(a lot).

**Comment by dantol29** [comment]
@552020 Sorry, I did git rebase. Commit history is a bit messed up

**Comment by 552020** [comment]
> @552020 Sorry, I did git rebase. Commit history is a bit messed up

Not a problem. I did finally some actual work, but I'm not sure how to test it. The backend tests the code in production right? @dantol29

**Comment by 552020** [comment]
@dantol29 I had to risolve a lot of conflicts. It seemed to me that in your rebase you just rebased without changing the code, so I just accepted my version.

**Comment by dantol29** [comment]
> > @552020 Sorry, I did git rebase. Commit history is a bit messed up
> 
> Not a problem. I did finally some actual work, but I'm not sure how to test it. The backend tests the code in production right? @dantol29

Run it locally and test with postman

> @dantol29 I had to risolve a lot of conflicts. It seemed to me that in your rebase you just rebased without changing the code, so I just accepted my version.

I changed the author of the commits. There was my work email

File: src/routes/chat_route.py [hunk 1]
```diff
@@ -192,7 +189,47 @@ async def add_user_to_agreed():
...
+                    continue
+
+                client = TelegramClient(
+                    StringSession(user_db_session.id), API_ID, API_HASH
+                )
+                if await connect_client(client, None, user.id) == -1:
+                    print("Error in connecting to Telegram")
+                    # we still return 200, because the chat is sold
+                    continue
+                if not await client.is_user_authorized():
```
Comment (dantol29) [inline src/routes/chat_route.py]: We have to disconnect from Telegram after each successful connect. Maybe add it under `if not await client.is_user_authorized():` so we always disconnect

**Review by dantol29 — COMMENTED** [review]
Great job! 🥇  Now code is indeed polished and beautiful. Would be great if you could you test your PR with postman in order to be sure that every endpoint that you changed works fine.
BTW, can you share your formatting settings?

**Comment by 552020** [comment]
> Great job! 🥇 Now code is indeed polished and beautiful. Would be great if you could you test your PR with postman in order to be sure that every endpoint that you changed works fine. BTW, can you share your formatting settings?

Look at this @lmangall! I got a compliment from our teach lead! I deserver a pay raise!

**Comment by dantol29** [comment]
> > Great job! 🥇 Now code is indeed polished and beautiful. Would be great if you could you test your PR with postman in order to be sure that every endpoint that you changed works fine. BTW, can you share your formatting settings?
> 
> Look at this @lmangall! I got a compliment from our teach lead! I deserver a pay raise!

I agree! You deserve it! Multiply your salary by 2!

**Comment by dantol29** [comment]
@552020 In `/add-user-to-agreed` we want to check a session for each user of the chat until we find an active session. This is just a for loop with `session.query(Session).filter(Session.user_id == str(userId)).one()` inside.

**Comment by dantol29** [comment]
@552020 I always get this error: 
```
{
    "error": "'sessionmaker' object does not support the asynchronous context manager protocol"
}
```

**Comment by dantol29** [comment]
Just Wow! To be honest I am afraid to merge, but we have to implement /quests ASAP

**Pull request merged**